mavros_extras/src/plugins/landing_target.cpp

```diff
@@ -0,0 +1,198 @@
+#include <geometry_msgs/PoseWithCovarianceStamped.h>
+
+namespace mavplugin {
+/**
+ * @brief Landing Target plugin
+ *
+ * This plugin is intended to publish the location of a landing area captured from a downward facing camera
+ * to the FCU and/or receive IRLock infrared tracking data.
+ */
+class LandingTargetPlugin : public MavRosPlugin,
+	private TFListenerMixin<LandingTargetPlugin> {
+public:
+	LandingTargetPlugin() :
+		sp_nh("~landing_target"),
+		uas(nullptr),
+		tf_rate(10.0),
+		send_tf(true),
+		listen_tf(false)
+	{ };
+
+	void initialize(UAS &uas_)
+	{
+		uas = &uas_;
+
+		sp_nh.param("send_tf", send_tf, true);
+		sp_nh.param("listen_tf", listen_tf, false);
+		sp_nh.param<std::string>("frame_id", frame_id, "landing_target");
+		sp_nh.param<std::string>("child_frame_id", child_frame_id, "camera_center");
+		sp_nh.param("tf_rate_limit", tf_rate, 50.0);
+
+		land_target_pub = sp_nh.advertise<geometry_msgs::PoseStamped>("landing_target", 10);
+
+		if (listen_tf) {
+			ROS_INFO_STREAM_NAMED("landing_target", "Listen to landing_target transform " << frame_id
+											<< " -> " << child_frame_id);
+			tf_start("LandingTargetTF", &LandingTargetPlugin::send_landing_target);
+		}
+
+		land_target_sub = sp_nh.subscribe("target", 10, &LandingTargetPlugin::land_target_cb, this);
+		local_position_sub = sp_nh.subscribe("/mavros/position/local", 10, &LandingTargetPlugin::local_pos_cb, this);
+	}
+
+	const message_map get_rx_handlers() {
+		return {
+			       MESSAGE_HANDLER(MAVLINK_MSG_ID_LANDING_TARGET, &LandingTargetPlugin::handle_landing_target)
+		};
+	}
+
+private:
+	friend class TFListenerMixin;
+	ros::NodeHandle sp_nh;
+	UAS *uas;
+
+	bool send_tf;
+	bool listen_tf;
+	double tf_rate;
+	ros::Time last_transform_stamp;
+	tf::TransformBroadcaster tf_broadcaster;
+
+	tf::Transform transform;
+	ros::Time lp_time;
+
+	std::string frame_id;
+	std::string child_frame_id;
+
+	ros::Publisher land_target_pub;
+	ros::Subscriber land_target_sub;
+	ros::Subscriber local_position_sub;
+
+	/* -*- low-level send -*- */
+
+	void landing_target(uint8_t target_num,
+			uint8_t frame,
+			float angle_x,
+			float angle_y,
+			float distance) {
+		mavlink_message_t msg;
+		mavlink_msg_landing_target_pack_chan(UAS_PACK_CHAN(uas), &msg,
+				target_num,
+				frame,
+				angle_x,
+				angle_y,
+				distance);
+		UAS_FCU(uas)->send_message(&msg);
+	}
+
+	/* -*- mid-level helpers -*- */
+
+	/*	AUXILIAR MATH:
+	*	d = sqrt(x^2 + y^2 + z^2)
+	*	x = sqrt((d^2-z^2) * cos^2(ax))
+	*	y = x*tan(ax)
+	*	z = d^2 - x^2 - y^2;
+	*/
+
+	/**
+	 * Send landing target transform to FCU
+	 */
+	void send_landing_target(const tf::Transform &transf, const ros::Time &stamp) {
+		// origin position in ROS ENU frame
+		tf::Vector3 pos = transf.getOrigin();
+
+		float ang_x = atan(pos.y()/pos.x());
+		float ang_y = boost::math::constants::pi<double>()/2 - atan(pos.y()/pos.x());
+		float distance = sqrt(pow(pos.x(),2) + pow(pos.y(),2) + pow(pos.z(),2));
+
```

Comment:
@LorenzMeier the above represent the calculations of X,Y offset angles and distance to target based on the target position. This was the way I see we can get this values, based on simple trigonometry.